spec/domain/accounts/users_spec.exs

```diff
@@ -4,130 +4,164 @@ defmodule Squadster.Domain.Accounts.UsersSpec do
 
   alias Squadster.Accounts.User
 
-  describe "changeset" do
-    context "when params are valid" do
-      it "is valid" do
-        %{valid?: is_valid} = %{first_name: "John", last_name: "Galt"} |> User.changeset
-        expect is_valid |> to(be true)
+  describe "changeset functions" do
+    defmodule ItSetsHashId do
+      use ESpec, shared: true
+
+      let_overridable :changeset_function
+
+      context "when user does not have hash_id yet" do
+        let :user, do: insert(:user, hash_id: nil)
+
+        it "should set hash_id" do
+          %{changes: changes} = User |> apply(changeset_function(), [user(), %{}])
+          expect changes
+          |> Map.has_key?(:hash_id)
+          |> to(be true)
+        end
       end
-    end
 
-    context "when params are invalid" do
-      context "when first_name is not set" do
-        it "is not valid" do
-          %{valid?: is_valid} = %{last_name: "Galt"} |> User.changeset
-          expect is_valid |> to(be false)
+      context "when user has hash_id" do
+        let :user, do: insert(:user)
+
+        it "should not update hash_id" do
+          %{changes: changes} = User |> apply(changeset_function(), [user(), %{}])
+          expect changes
+          |> Map.has_key?(:hash_id)
+          |> to(be false)
         end
       end
+    end
 
-      context "when last_name is not set" do
-        it "is not valid" do
-          %{valid?: is_valid} = %{first_name: "John"} |> User.changeset
-          expect is_valid |> to(be false)
+    describe "changeset" do
+      context "when params are valid" do
+        it "is valid" do
+          %{valid?: is_valid} = %{first_name: "John", last_name: "Galt"} |> User.changeset
+          expect is_valid |> to(be true)
         end
+
+        it_behaves_like ItSetsHashId, changeset_function: :changeset
       end
 
-      context "when mobile_phone contains illegal symbols" do
-        it "is not valid" do
-          %{valid?: is_valid} = %{first_name: "John", last_name: "Galt", mobile_phone: "&123"} |> User.changeset
-          expect is_valid |> to(be false)
+      context "when params are invalid" do
+        context "when first_name is not set" do
+          it "is not valid" do
+            %{valid?: is_valid} = %{last_name: "Galt"} |> User.changeset
+            expect is_valid |> to(be false)
+          end
+        end
+
+        context "when last_name is not set" do
+          it "is not valid" do
+            %{valid?: is_valid} = %{first_name: "John"} |> User.changeset
+            expect is_valid |> to(be false)
+          end
+        end
+
+        context "when mobile_phone contains illegal symbols" do
+          it "is not valid" do
+            %{valid?: is_valid} = %{first_name: "John", last_name: "Galt", mobile_phone: "&123"} |> User.changeset
```

Comment:
Let's move this to a variable. It will help us to understand, where are valid params, and where are not